// @tsed/common/lib/platform/domain/PlatformContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestContext = exports.PlatformContext = void 0;
const di_1 = require("@tsed/di");
const RequestLogger_1 = require("./RequestLogger");
class PlatformContext extends Map {
    constructor({ id, injector, logger, response, request, endpoint, ...options }) {
        super();
        /**
         * Date when request have been handled by the server. @@RequestLogger@@ use this date to log request duration.
         */
        this.dateStart = new Date();
        /**
         * The request container used by the Ts.ED DI. It contain all services annotated with `@Scope(ProviderScope.REQUEST)`
         */
        this.container = new di_1.LocalsContainer();
        this.id = id;
        injector && (this.injector = injector);
        response && (this.response = response);
        request && (this.request = request);
        endpoint && (this.endpoint = endpoint);
        this.logger = new RequestLogger_1.RequestLogger(logger, {
            ...options,
            id,
            dateStart: this.dateStart
        });
    }
    get env() {
        return this.injector.settings.env;
    }
    async destroy() {
        await this.container.destroy();
        this.logger.destroy();
        this.response.destroy();
        this.request.destroy();
        // @ts-ignore
        delete this.container;
        // @ts-ignore
        delete this.logger;
        // @ts-ignore
        delete this.injector;
        // @ts-ignore
        delete this.endpoint;
        // @ts-ignore
        delete this.response;
        // @ts-ignore
        delete this.request;
    }
    async emit(eventName, ...args) {
        return this.injector && this.injector.emit(eventName, ...args);
    }
    /**
     * Return the original request instance.
     */
    getRequest() {
        return this.request.raw;
    }
    /**
     * Return the original response instance.
     */
    getResponse() {
        return this.response.raw;
    }
}
exports.PlatformContext = PlatformContext;
/**
 * @deprecated Use PlatformContext instead.
 */
class RequestContext extends PlatformContext {
}
exports.RequestContext = RequestContext;
